sdk/pinocchio/src/sysvars/instructions.rs

```diff
@@ -0,0 +1,236 @@
+};
+
+use core::{marker::PhantomData, mem::size_of, ops::Deref};
+
+/// Sysvar1nstructions1111111111111111111111111
+pub const INSTRUCTIONS_ID: Pubkey = [
+    0x06, 0xa7, 0xd5, 0x17, 0x18, 0x7b, 0xd1, 0x66, 0x35, 0xda, 0xd4, 0x04, 0x55, 0xfd, 0xc2, 0xc0,
+    0xc1, 0x24, 0xc6, 0x8f, 0x21, 0x56, 0x75, 0xa5, 0xdb, 0xba, 0xcb, 0x5f, 0x08, 0x00, 0x00, 0x00,
+];
+
+pub struct Instructions<T>
+where
+    T: Deref<Target = [u8]>,
+{
+    data: T,
+}
+
+impl<T> Instructions<T>
+where
+    T: Deref<Target = [u8]>,
+{
+    /// Creates a new `Instructions` struct.
+    ///
+    /// `data` is the instructions sysvar account data.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check if the provided data is from the Sysvar Account.
+    pub unsafe fn new(data: T) -> Self {
+        Instructions { data }
+    }
+
+    /// Load the current `Instruction`'s index in the currently executing
+    /// `Transaction`.
+    pub fn load_current_index(&self) -> u16 {
+        unsafe {
+            let len = self.data.len();
+            u16::from_le_bytes(*(self.data.as_ptr().add(len - 2) as *const [u8; 2]))
+        }
+    }
+
+    /// Creates and returns an `IntrospectedInstruction` for the instruction at the specified index.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check if the provided index is out of bounds. It is
+    /// typically used internally with the `load_instruction_at` or `get_instruction_relative` functions,
+    /// which perform the necessary index verification.
+    pub unsafe fn deserialize_instruction(&self, index: usize) -> IntrospectedInstruction {
+        let offset = u16::from_le(
+            *(self
+                .data
+                .as_ptr()
+                .add(size_of::<u16>() + index * size_of::<u16>()) as *const u16),
+        );
+
+        IntrospectedInstruction {
+            raw: self.data.as_ptr().add(offset as usize),
+            marker: PhantomData,
+        }
+    }
+
+    /// Creates and returns an `IntrospectedInstruction` for the instruction at the specified index.
+    #[inline(always)]
+    pub fn load_instruction_at(
+        &self,
+        index: usize,
+    ) -> Result<IntrospectedInstruction, ProgramError> {
+        unsafe {
+            let num_instructions = u16::from_le(*(self.data.as_ptr() as *const u16));
+            if index >= num_instructions as usize {
+                return Err(SanitizeError::IndexOutOfBounds)
+                    .map_err(|_| ProgramError::InvalidInstructionData);
+            }
+
+            Ok(self.deserialize_instruction(index))
+        }
+    }
+
+    /// Creates and returns an `IntrospectedInstruction` relative to the current `Instruction` in the
+    /// currently executing `Transaction.
+    #[inline(always)]
+    pub fn get_instruction_relative(
+        &self,
+        index_relative_to_current: i64,
+    ) -> Result<IntrospectedInstruction, ProgramError> {
+        unsafe {
```

Comment:
Same comments for the "unsafe", `u16::to_le` and `ProgramError` things. Also it contains duplicated logic with `load_instruction_at` so we can just call that method instead.
```rust
let current_index = self.load_current_index() as i64;
let index = current_index.saturating_add(index_relative_to_current);

if index < 0 {
    return Err(ProgramError::InvalidArgument);
}

self.load_instruction_at(index as usize)
```